### PR TITLE
Archive Incidents and waypoints

### LINKED ISSUE #58: archive incidents and waypoints
### Task:
- [x] Delete incidents that have already occurred (incident_end has passed).
- [x] Archive waypoints into the performance_archive.

### User Story:
As a user, I want old data to not be visible on the dashboard and alerts so that I can view up to date data.

### PR DESCRIPTION
**Pull request by FarihaChoudhury**
- [x] Cleans database of out of date incidents.
- [x] Cleans database and archives out of date waypoints, and removes connected cancellations. 
- [x] Lambda for running archive process.
- [x] Dockerfile for dockerising archive lambda

closes #58

### CODE REVIEW

File: archive/README.md
Comment (A-Faris) [inline archive/README.md:1]: Include how to create a docker image

File: archive/archive.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    logging.info("Clean: Began cleaning process.")
+
+    nr_cleaner.clean_national_rail_incidents()
+
+    rtt_cleaner.clean_real_time_trains_data()
+
+    logging.info("Clean: Completed cleaning process.")
+
+
+def handler(event, context):
```
Comment (A-Faris) [inline archive/archive.py:23]:
```suggestion
def handler(_event, _context):
```

File: archive/archive.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    # pylint: disable=unused-argument
```
Comment (A-Faris) [inline archive/archive.py]:
```suggestion
def handler(_event, _context):
```

File: archive/archive.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    logging.basicConfig(level=logging.INFO,
+                        format="%(asctime)s - %(levelname)s - %(message)s")
+
+    clean_rail_tracker()
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO,
+                        format="%(asctime)s - %(levelname)s - %(message)s")
+    clean_rail_tracker()
```
Comment (A-Faris) [inline archive/archive.py]:
```suggestion
    handler(None, None)
```

File: archive/clean_national_rail.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+""" This module works with old NationalRail incident data from the RDS and 
+    deletes incident data that is no longer needed."""
+
+import logging
+
+from psycopg2.extensions import connection
+import db_connection
```
Comment (A-Faris) [inline archive/clean_national_rail.py]:
```suggestion
from db_connection import execute_without_result, get_connection
```

File: archive/db_connection.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+def get_cursor(conn: connection) -> cursor:
+    """ Retrieves cursor and returns it. """
+    return conn.cursor(cursor_factory=RealDictCursor)
+
+
+def execute(conn: connection, query: str, data: tuple) -> dict | None:
+    """ Executes SQL queries on AWS RDS, and returns result.
+        Uses fetchall(). """
+
+    query_command = query.strip().split(" ")[0]
```
Comment (A-Faris) [inline archive/db_connection.py:33]:
```suggestion
    query_command = query.strip().split()[0]
```

File: archive/DOCKERFILE
Comment (A-Faris) [inline archive/DOCKERFILE:1]: Why are you shouting this filename?

File: archive/clean_national_rail.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+def delete_old_incidents(conn: connection) -> None:
+    """ Deletes incidents with end dates that have passed and are no longer required 
+        for storage. This also deletes affected operators (ON DELETE CASCADE)"""
+
+    query = """
+        DELETE FROM incident
+        WHERE incident_end < TIMEZONE('Europe/London', CURRENT_TIMESTAMP);
+    """
+
+    db_connection.execute_without_result(conn, query, ())
```
Comment (A-Faris) [inline archive/clean_national_rail.py]:
```suggestion
    execute_without_result(conn, query, ())
```

File: archive/clean_national_rail.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def clean_national_rail_incidents() -> None:
+    """ Cleans NationalRail Incidents from the RDS based on whether the incident has ended 
+        or not."""
+
+    conn = db_connection.get_connection()
```
Comment (A-Faris) [inline archive/clean_national_rail.py]:
```suggestion
    conn = get_connection()
```

File: archive/clean_national_rail.py
Comment (A-Faris) [inline archive/clean_national_rail.py:1]: Consider reducing to one function

File: archive/clean_real_time_trains.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
+""" This module works with old RealTimeTrains waypoint data from the RDS and 
+    archives train data that is no longer needed. The archive contains the 
+    performance metrics of the train data."""
+
+import logging
+
+from psycopg2.extensions import connection
+import db_connection
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
from db_connection execute, execute_without_result
```

File: archive/clean_real_time_trains.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+def get_all_stations(conn: connection) -> list[dict]:
+    """ Retrieves all waypoints that have a run_date of a month ago or more. """
+
+    query = """
+        SELECT station_id
+        FROM station;
+    """
+
+    station_ids = db_connection.execute(conn, query, ())
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    station_ids = execute(conn, query, ())
```

File: archive/clean_real_time_trains.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    """ Retrieves all waypoints that have a run_date of a month ago or more. """
+
+    query = """
+        SELECT *
+        FROM waypoint
+        WHERE run_date <= CURRENT_DATE - INTERVAL '1 month' 
+            AND station_id = %s;
+    """
+
+    outdated_waypoints = db_connection.execute(conn, query, (station_id,))
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    outdated_waypoints = execute(conn, query, (station_id,))
```

File: archive/clean_real_time_trains.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+def get_month_old_waypoints(conn: connection, station_id: str) -> list[dict]:
+    """ Retrieves all waypoints that have a run_date of a month ago or more. """
+
+    query = """
+        SELECT *
+        FROM waypoint
+        WHERE run_date <= CURRENT_DATE - INTERVAL '1 month' 
+            AND station_id = %s;
+    """
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py:32]:
```suggestion
    query = """SELECT * FROM waypoint
        WHERE run_date <= CURRENT_DATE - INTERVAL '1 month' 
            AND station_id = %s;"""
```

File: archive/db_connection.py (hunk 5 above)
Comment (FarihaChoudhury) [reply archive/db_connection.py:33]: Nice ill include it

File: archive/clean_real_time_trains.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        WHERE station_id = %s
+            AND run_date <= CURRENT_DATE - INTERVAL '1 month'
+        GROUP BY station_id;
+    """
+
+    avg_delay = db_connection.execute(conn, query, (station_id,))
+
+    if avg_delay:
+        avg_delay = avg_delay[0]['avg_overall_delay']
+    return avg_delay
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    return avg_delay[0]['avg_overall_delay'] if avg_delay else None
```

File: archive/clean_real_time_trains.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+        SELECT 
+            station_id,
+            AVG((actual_arrival - booked_arrival)+(actual_departure - booked_departure))avg_overall_delay
+        FROM waypoint
+        WHERE station_id = %s
+            AND run_date <= CURRENT_DATE - INTERVAL '1 month'
+        GROUP BY station_id;
+    """
+
+    avg_delay = db_connection.execute(conn, query, (station_id,))
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    avg_delay = execute(conn, query, (station_id,))
```

File: archive/README.md
Comment (FarihaChoudhury) [reply archive/README.md:1]: Good idea

File: archive/archive.py (hunk 2 above)
Comment (FarihaChoudhury) [reply archive/archive.py]: Good idea

File: archive/clean_real_time_trains.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    return avg_delay
+
+
+def compute_cancellation_count_for_station(conn: connection, station_id: int) -> dict:
+    """ Computes and returns the number of cancellations for a station, for trains that 
+        have arrived/ departed over a month ago. """
+
+    query = """
+        SELECT 
+            COUNT(cancellation_id) AS cancellation_count
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
        SELECT COUNT(cancellation_id) AS cancellation_count
```

File: archive/archive.py (hunk 3 above)
Comment (FarihaChoudhury) [reply archive/archive.py]: Nice

File: archive/DOCKERFILE
Comment (FarihaChoudhury) [reply archive/DOCKERFILE:1]: typo

File: archive/clean_real_time_trains.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+        WHERE w.station_id = %s
+            AND run_date <= CURRENT_DATE - INTERVAL '1 month';
+    """
+
+    cancellation_count = db_connection.execute(conn, query, (station_id,))
+
+    if cancellation_count:
+        cancellation_count = cancellation_count[0]['cancellation_count']
+
+    return cancellation_count
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    return cancellation_count[0]['cancellation_count'] if cancellation_count else None
```

File: archive/clean_real_time_trains.py [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+        SELECT 
+            COUNT(cancellation_id) AS cancellation_count
+        FROM waypoint w
+        JOIN cancellation c 
+        ON c.waypoint_id = w.waypoint_id
+        WHERE w.station_id = %s
+            AND run_date <= CURRENT_DATE - INTERVAL '1 month';
+    """
+
+    cancellation_count = db_connection.execute(conn, query, (station_id,))
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    cancellation_count = execute(conn, query, (station_id,))
```

File: archive/clean_real_time_trains.py [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+def insert_performance_archive(conn: connection, archive_data: dict) -> None:
+    """ Insert performance data about each station into the archive.
+        Creation date is date of inserting into archive. """
+
+    query = """
+        INSERT INTO performance_archive (station_id, avg_delay, cancellation_count, creation_date) 
+        VALUES (%s, %s, %s, TIMEZONE('Europe/London', CURRENT_TIMESTAMP));
+    """
+
+    db_connection.execute_without_result(conn, query, (archive_data["station_id"],
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    execute_without_result(conn, query, (archive_data["station_id"],
```

File: archive/clean_national_rail.py
Comment (FarihaChoudhury) [reply archive/clean_national_rail.py:1]: yes good idea

File: archive/clean_real_time_trains.py (hunk 11 above)
Comment (FarihaChoudhury) [reply archive/clean_real_time_trains.py:32]: I like it spaced because its readable

File: archive/clean_real_time_trains.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def delete_cancellation(conn: connection, waypoint_id: int) -> None:
+    """ Deletes a cancellation record by its waypoint id. """
+
+    query = """
+        DELETE FROM cancellation 
+        WHERE waypoint_id = %s;
+    """
+    db_connection.execute_without_result(conn, query, (waypoint_id,))
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    execute_without_result(conn, query, (waypoint_id,))
```

File: archive/clean_real_time_trains.py [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def delete_waypoint(conn: connection, waypoint_id: int) -> None:
+    """ Deletes a waypoint record by its waypoint id. """
+
+    query = """
+        DELETE FROM waypoint 
+        WHERE waypoint_id = %s;
+    """
+    db_connection.execute_without_result(conn, query, (waypoint_id,))
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    execute_without_result(conn, query, (waypoint_id,))
```

File: archive/clean_real_time_trains.py [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def get_table_size(conn: connection, table_name: str) -> int:
+    """ Returns the size of the table given by the table_name argument. """
+
+    query = f"""
+        SELECT COUNT(*) FROM {table_name};
+    """
+    return db_connection.execute(conn, query, ())
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    return execute(conn, query, ())
```

File: archive/clean_real_time_trains.py [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def clean_real_time_trains_data():
+    """ Cleans RealTimeTrains waypoints data from RDS based on how long ago the train journey
+        occurred. Computes performance statistics for the archiving process and inserts
+        into archive.  """
+
+    conn = db_connection.get_connection()
+
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py]:
```suggestion
    with get_connection() as conn:

```

File: database/query.sql
Comment (A-Faris) [inline database/query.sql:1]: Why is everything commented out?

File: archive/db_connection.py [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+
+    query_command = query.strip().split()[0]
+
+    try:
+        cur = get_cursor(conn)
+
+        cur.execute(query, (data))
+
+        conn.commit()
+        cur.close()
```
Comment (A-Faris) [inline archive/db_connection.py]:
```suggestion
        with get_cursor(conn) as cur:
            cur.execute(query, (data))
            conn.commit()
```

File: archive/db_connection.py [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+    query_command = query.strip().split(" ")[0]
+
+    try:
+        cur = get_cursor(conn)
+
+        cur.execute(query, (data))
+
+        conn.commit()
+        result = cur.fetchall()
+        cur.close()
```
Comment (A-Faris) [inline archive/db_connection.py]:
```suggestion
        with get_cursor(conn) as cur:
            cur.execute(query, (data))
            conn.commit()
            result = cur.fetchall()
```

File: database/query.sql
Comment (FarihaChoudhury) [reply database/query.sql:1]: idk

File: archive/clean_real_time_trains.py [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    execute_without_result(conn, query, (waypoint_id,))
+
+
+def delete_waypoint(conn: connection, waypoint_id: int) -> None:
+    """ Deletes a waypoint record by its waypoint id. """
+
+    query = """
+        DELETE FROM waypoint 
+        WHERE waypoint_id = %s;
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py:105]:
```suggestion
        WHERE waypoint_id = %s,
        RETURNING *;
```

File: archive/clean_real_time_trains.py [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    execute_without_result(conn, query, (waypoint_id,))
+
+
+def get_table_size(conn: connection, table_name: str) -> int:
+    """ Returns the size of the table given by the table_name argument. """
+
+    query = f"""
+        SELECT COUNT(*) FROM {table_name};
+    """
```
Comment (A-Faris) [inline archive/clean_real_time_trains.py:115]:
```suggestion
    query = f"SELECT COUNT(*) FROM {table_name};"
```

**Review by A-Faris — APPROVED** [review]
Yep all good

**Pull request merged**